### PR TITLE
trie: pass TestDerivableList

### PR DESCRIPTION
**Pull request by gballet**
These are the changes needed for `TestDerivableList` to pass:

* include the required change for `rawNode` to be serialized correctly
* alternative to `force`: make sure that the root of a trie is hashed if its length is less than 32 bytes
* remove one of @holiman's alloc optimization because it doesn't seem to work properly, have to investigate why

### CODE REVIEW

File: trie/database.go [hunk 1]
```diff
@@ -99,6 +99,15 @@ type rawNode []byte
...
 
+func (n rawNode) EncodeRLP(w io.Writer) error {
+	count, err := w.Write([]byte(n))
+	if err != nil && count != len(n) {
+		err = fmt.Errorf("rawNode: could not write full length")
+	}
+
+	return err
+}
+
```
Comment (holiman) [inline trie/database.go:110]: Why is this needed?

Comment (gballet) [reply trie/database.go:110]: because otherwise it will add a `0x80+size` prefix when including it into the parent, which is redundant with the one already present in the `rawNode` at this stage. This in turn gives the wrong hash value for the parent. It only occurs when the RLP size is < 32

Comment (holiman) [reply trie/database.go:110]: But.. what makes the `w.Write` fail? And then you just wrap it in another error (using `Errorf` but no param to it)?

File: trie/stacktrie.go [hunk 2]
```diff
@@ -363,6 +361,15 @@ func (st *StackTrie) hash() []byte {
...
 	st.val = st.hash()
 	st.nodeType = hashedNode
+	if len(st.val) < 32 {
+		ret := make([]byte, 32)
+		h := newHasher(false)
+		defer returnHasherToPool(h)
+		h.sha.Reset()
+		h.sha.Write(h.tmp)
+		h.sha.Read(ret)
+		return common.BytesToHash(ret)
```
Comment (holiman) [inline trie/stacktrie.go:371]: shouldn't you forcibly set the `st.val` to the forced hash aswell? 
also, does this somehow screw up Commit (hashing a preimage but not sending it out for Commit)

File: trie/stacktrie.go [hunk 3]
```diff
@@ -363,6 +364,15 @@ func (st *StackTrie) hash() []byte {
 func (st *StackTrie) Hash() (h common.Hash) {
 	st.val = st.hash()
 	st.nodeType = hashedNode
+	if len(st.val) < 32 {
+		ret := make([]byte, 32)
+		h := newHasher(false)
+		defer returnHasherToPool(h)
+		h.sha.Reset()
+		h.sha.Write(h.tmp)
```
Comment (holiman) [inline trie/stacktrie.go:372]: Horrifically nasty bug here! This _works_, but it's _damn_ brittle! You're writing `h.tmp` from a `h` just retrieved from the pool. And it just so happens that it contains what we want it to contain,  since we just returned it to the pool and didn't clear it. Should be 
```
		// If the 'st.val' is < 32, it's not the hash, but the preimage
		// (the  rlp-encoding of the node). For the top level node, we need
		// to force the hashing
		ret := make([]byte, 32)
		h := newHasher(false)
		defer returnHasherToPool(h)
		h.sha.Reset()
		h.sha.Write(st.val)
```

```suggestion
		h.sha.Write(st.val)
```

Comment (gballet) [reply trie/stacktrie.go:372]: oh wow good catch

**Comment by holiman** [comment]
lgtm. I have som new changes too. Should I push them on this, or do you want to merge this first and I can make yet another PR?

With my changes:  
```
name                       old time/op    new time/op    delta
DeriveSha200/std_trie-6       576µs ±13%     567µs ±15%     ~     (p=0.690 n=5+5)
DeriveSha200/stack_trie-6     509µs ±16%     468µs ± 3%   -7.99%  (p=0.008 n=5+5)

name                       old alloc/op   new alloc/op   delta
DeriveSha200/std_trie-6       275kB ± 0%     275kB ± 0%     ~     (p=0.595 n=5+5)
DeriveSha200/stack_trie-6    60.9kB ± 0%    55.0kB ± 0%   -9.69%  (p=0.008 n=5+5)

name                       old allocs/op  new allocs/op  delta
DeriveSha200/std_trie-6       2.87k ± 0%     2.87k ± 0%     ~     (all equal)
DeriveSha200/stack_trie-6     1.65k ± 0%     1.27k ± 0%  -23.24%  (p=0.008 n=5+5)

```

**Comment by gballet** [comment]
sure, please push

**Comment by gballet** [comment]
The mem profile looks good on a fast sync

profile001

**Comment by holiman** [comment]
Oh, I'm not allowed to push to your branch of a branch, it seems...

**Comment by holiman** [comment]
Let's merge this already..

Fast-sync done

Screenshot_2020-09-07 Dual Geth - Grafana(1)


Mem consumption same same
Allocs markedly less with stacktrie, but OTOH it also imported blocks at a slower rate, so hard to say a lot about that. It definitely doesn't seem to bork anything, at least :)

File: trie/database.go [hunk 4]
```diff
@@ -99,6 +99,15 @@ type rawNode []byte
 func (n rawNode) cache() (hashNode, bool)   { panic("this should never end up in a live trie") }
 func (n rawNode) fstring(ind string) string { panic("this should never end up in a live trie") }
 
+func (n rawNode) EncodeRLP(w io.Writer) error {
+	_, err := w.Write([]byte(n))
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (holiman) [inline trie/database.go:108]: How about just 
```suggestion
	_, err := w.Write([]byte(n))
	return err
```

This change does modify a struct that's already used by the trie today, and apparently in a way which modifies the behaviour of `rawNode`. So, I'm curious why this works, but even more curious as to why this change won't break older code, supposedly relying on how it worked before... ?

**Comment by gballet** [comment]
nice!

yeah I'll apply your last suggestion and I'm good for the merge. It's already one branch removed from making it into the main repo

File: trie/database.go (hunk 4 above)
Comment (gballet) [reply trie/database.go]: no "older" code was serializing `rawNode`, so no older code should be broken as far as I can tell
